src/agents/supervisor/agent.py

```diff
@@ -1,91 +1,298 @@
 
-    def __init__(self, model: IModel, members: list[str]):
-        self.model = model
-        self.options = [FINALIZER] + members
-        self.parser = self._create_parser()
-        self.supervisor_chain = self._create_supervisor_chain()
+        self.model = gpt_4o
+        self.members = members
+        self.options = members
+        self.parser = self._route_create_parser()
+        self.supervisor_chain = self._create_supervisor_chain(gpt_4o)
+        self._planner_chain = self._create_planner_chain(gpt_4o)
+        self._graph = self._build_graph()
 
-    def _create_parser(self) -> PydanticOutputParser:
+    def _get_members_str(self) -> str:
+        return ", ".join(self.members)
+
+    def _route_create_parser(self) -> PydanticOutputParser:
         class RouteResponse(BaseModel):
             next: Literal[*self.options] | None = Field(  # type: ignore
                 description="next agent to be called"
             )
 
         return PydanticOutputParser(pydantic_object=RouteResponse)
 
-    def _create_supervisor_chain(self) -> RunnableSequence:
+    def _create_supervisor_chain(self, model: IModel) -> RunnableSequence:
         supervisor_system_prompt = ChatPromptTemplate.from_messages(
             [
-                ("system", SUPERVISOR_ROLE_PROMPT),
+                ("system", ROUTER_ROLE_PROMPT),
                 MessagesPlaceholder(variable_name="messages"),
                 ("assistant", "Subtasks: {subtasks}"),
-                ("system", SUPERVISOR_TASK_PROMPT),
+                ("system", ROUTER_TASK_PROMPT),
             ]
         ).partial(
             options=str(self.options),
-            members=", ".join(self.options),
-            finalizer=FINALIZER,
+            members=self._get_members_str(),
+            end=str(END),
             output_format=self.parser.get_format_instructions(),
         )
 
-        return supervisor_system_prompt | self.model.llm  # type: ignore
+        return supervisor_system_prompt | model.llm  # type: ignore
 
     @property
     def name(self) -> str:
         """Agent name."""
         return self._name
 
-    def agent_node(self):  # noqa ANN
+    def agent_node(self) -> CompiledGraph:
         """Get Supervisor agent node function."""
+        return self._graph
+
+    def _route(self, state: AgentState) -> dict[str, Any]:
+        """Supervisor node."""
+        try:
+            result = self.supervisor_chain.invoke(
+                input={
+                    "messages": filter_messages(state.messages),
+                    "subtasks": json.dumps(
+                        [subtask.__dict__ for subtask in state.subtasks]
+                    ),
+                },
+            )
+            route_result = self.parser.parse(result.content)
+            return {
+                "next": route_result.next,
+                "subtasks": state.subtasks,
+                "messages": [AIMessage(content=result.content, name=self.name)],
+            }
```

Comment:
Can this same thing be done without using LLM? Cann't we loop over the subtasks and check for next pending subtask? If all subtasks are marked as completed, then route to finalizer?

Reply:
you are right, done